src/_http/http2Adapter.js

```diff
@@ -0,0 +1,226 @@
+'use strict'
+var http2 = require('http2')
+var errors = require('./errors')
+var util = require('../_util')
+
+// Destroy session after 1 minute of inactivity.
+var DESTROY_HTTP2_SESSION_TIME = 1000 * 60
+var STREAM_PREFIX = 'stream::'
+
+/**
+ * Http client adapter built around NodeJS http2 module.
+ *
+ * @constructor
+ * @private
+ */
+function Http2Adapter() {
+  /**
+   * Identifies a type of adapter.
+   *
+   * @type {string}
+   */
+  this.type = 'http2'
+  /**
+   * Serves for reusing http2 sessions between multiple calls.
+   *
+   * @type {object}
+   * @private
+   */
+  this._sessionMap = {}
+}
+
+/**
+ * Resolves ClientHttp2Session to be reused across multiple requests.
+ *
+ * @param {string} origin Request origin to connect to.
+ * @param {?boolean} isStreaming Whether it's a streaming request. A separate session
+ * is created for streaming requests to avoid shared resources with regular
+ * ones for the purpose of reliability.
+ * @returns {ClientHttp2Session} Http2 session.
+ */
+Http2Adapter.prototype._resolveSessionFor = function(origin, isStreaming) {
+  var sessionKey = isStreaming ? STREAM_PREFIX + origin : origin
+
+  if (!this._sessionMap[sessionKey]) {
+    var self = this
+
+    var cleanup = function() {
+      self._cleanupSessionFor(origin, isStreaming)
+    }
+
+    // Initializing http2 session.
+    this._sessionMap[sessionKey] = http2
+      .connect(origin)
+      .once('error', cleanup)
+      .once('goaway', cleanup)
+      // Destroys http2 session after specified time of inactivity
+      // and releases event loop.
+      .setTimeout(DESTROY_HTTP2_SESSION_TIME, cleanup)
```

Comment:
Q: what does "inactivity" mean? Is it lack of network traffic or lack of open open streams?

Reply:
Yeah, it's a lack of any network traffic (including non-streaming requests)

Reply:
I mean streams in the HTTP2 spec sense. Would this cause idle **fauna streams** to close after the section timeout?

Reply:
No, it wouldn't. When there's an ongoing http2 stream in the session, it won't be closed